src/pynwb/icephys.py

```diff
@@ -319,15 +338,513 @@ def __get_row_ids(self, sweep_number):
+            # if response.sweep_number != stimulus.sweep_number:
+            #     warnings.warn("sweep_number are usually expected to be the same for PatchClampSeries type "
+            #                   "stimulus and response pairs in an intracellular recording.")
+            if response.electrode != stimulus.electrode:
+                raise ValueError("electrodes are usually expected to be the same for PatchClampSeries type "
+                                 "stimulus and response pairs in an intracellular recording.")
+
+        # Compile the electrodes table data
+        electrodes = copy(popargs('electrode_metadata', kwargs))
+        if electrodes is None:
+            electrodes = {}
+        electrodes['electrode'] = electrode
+
+        # Compile the stimuli table data
+        stimuli = copy(popargs('stimulus_metadata', kwargs))
+        if stimuli is None:
+            stimuli = {}
+        stimuli['stimulus'] = (stimulus_start_index, stimulus_index_count, stimulus)
+
+        # Compile the reponses table data
+        responses = copy(popargs('response_metadata', kwargs))
+        if responses is None:
+            responses = {}
+        responses['response'] = (response_start_index, response_index_count, response)
+
+        _ = super().add_row(enforce_unique_id=True,
+                            electrodes=electrodes,
+                            responses=responses,
+                            stimuli=stimuli,
+                            **kwargs)
+        return len(self) - 1
+
+    @staticmethod
+    def __compute_index(start_index, index_count, time_series, name):
+        """
+        Internal helper function to compute the start_index and index_count
+        to use for the stimulus and response column
+
+        :param start_index: The start_index provided by the user
+        :param index_count: The index count provided by the user
+        :param time_series: The timeseries object to reference. May be None.
+        :param name: Name of the table. Used only to enhance error reporting
+
+        :raises IndexError: If index_count cannot be determined or start_index+index_count
+            are outside of the range of the timeseries.
+
+        :returns: A tuple of integers with the start_index and index_count to use
+        """
+        if time_series is None:
+            return -1, -1
+        start_index = -1 if start_index is None else start_index
+        start_index = start_index if start_index >= 0 else 0
+        num_samples = time_series.num_samples
+        index_count = -1 if index_count is None else index_count
+        index_count = (index_count
+                       if index_count >= 0
+                       else ((num_samples - start_index)
+                             if num_samples is not None
+                             else None))
+        if index_count is None:
+            raise IndexError("Invalid %s_index_count cannot be determined from %s data." % (name, name))
```

Comment:
This structure is difficult for me to read.
```suggestion
        if index_count is None or index_count < 0:
            if num_samples is not None:
                index_count = num_samples - start_index
            else:
                raise IndexError("Invalid %s_index_count cannot be determined from %s data." % (name, name))
```

Reply:
I simplified the logic in the latest commit and added comments.